### DISCUSSION TITLE
Multiple variable sub routes

### DISCUSSION THREAD

**Discussion by leoshusar** [csells/go_router#128] [Q&A] [answered]
Hi!
I have an app with TabView as a main screen.
Those views have routes `/(users|roles)`.
Now I would like to do something like this:
If you click on `edit user` button, go to `/users/:userId/(info|roles)`
and
If you click on `edit role` button, go to `/roles/:roleId/(info|actions)`
Both "edit routes" should be descendants of main screen.

however I cannot think of any way to achieve this.

```dart
GoRoute(
  path: '/:main(users|roles)',
  pageBuilder: (c, s) => MaterialPage<void>(
    child: MultiProvider(
      providers: [
        Provider<UsersBloc>(...),
        Provider<RolesBloc>(...),
      ],
      child: TabsScreen(),
    ),
  ),
  routes: [
    GoRoute(
      path: '', // somehow match /users/:userId/(info|roles)
    ),
    GoRoute(
      path: '', // somehow match /roles/:roleId/(info|roles)
    ),
  ],
)
```

Could you please help me if this is possible?

EDIT: I now see you have something like this in your books example, but I am also using provider + bloc for each tab view

**Comment by csells** [comment]
The data provider you use should not affect the navigation.

**Reply by leoshusar** [reply]
By looking at that books example I've got an idea I need to split it like
```dart
GoRoute(
  path: '/users', // tab users,
  pageBuilder: .. MainPage(tab: 'users'),
),
GoRoute(
  path: '/roles', // tab roles,
  pageBuilder: .. MainPage(tab: 'roles'),
),
```
but I believe this is not the option, because this then rerenders the whole page and doesn't act like TabView page change.

And when I use my original code example I don't know how to match subroute of regex route.
Something like this
```dart
GoRoute(
  path: '/:main(users|roles)',
  ),
  routes: [
    GoRoute( 
      path: '', // if ancestor is `users`, match /users/:userId/(info|roles),
      pageBuilder: ... UserInfoScreen(),
    ),
    GoRoute(
      path: '', // if ancestor is `roles`, match /roles/:roleId/(info|roles|something)
      pageBuilder: ... RoleInfoScreen(),
    ),
  ],
)
```

So when I open `/users` or `/roles`, I get `TabsScreen`.
When I open `/users/1/info`, I should get `UserInfoScreen`.
When I open `/roles/1/info`, I should get `RoleInfoScreen`.

I hope I explained it better now.

**Comment by csells** [comment, marked as answer]
You could arrange your routes this way:

```dart
GoRoute(
  path: '/users',
  pageBuilder: ...TabsScreen(main: 'users'),
  routes: [
    GoRoute( 
      path: ':userId/(info|roles),
      pageBuilder: ... UserInfoScreen(userId: state.params['userId']),
    ),
  ],
),
GoRoute(
  path: '/roles',
  pageBuilder: ...TabsScreen(main: 'roles'),
  routes: [
    GoRoute(
      path: ':roleId/(info|roles|something)',
      pageBuilder: ... RoleInfoScreen(roleId: state.params['roleId']!),
    ),
  ],
)
```

**Reply by leoshusar** [reply]
Thank you, I think I now know where my problem was.

I tried this configuration before but I have set it like this:
```dart
GoRoute(
  path: '/users', // tab users,
  pageBuilder: .. Provider<UsersBloc> -> TabsScreen(tab: 'users'),
),
GoRoute(
  path: '/roles', // tab roles,
  pageBuilder: .. Provider<RolesBloc> -> TabsScreen(tab: 'roles'),
),
```
and this didn't work because when I switched tabs from users to roles, it wanted `RolesBloc` before that `Provider<RolesBloc>` was created.

Now it works when I pass all blocs everytime:
```dart
GoRoute(
  path: '/users', // tab users,
  pageBuilder: .. MultiProvider[
    Provider<UsersBloc>,
    Provider<RolesBloc>,
  ] -> TabsScreen(tab: 'users'),
),
GoRoute(
  path: '/roles', // tab roles,
  pageBuilder: .. MultiProvider[
    Provider<UsersBloc>,
    Provider<RolesBloc>,
  ] -> TabsScreen(tab: 'roles'),
),
```

So my last question, I guess there is no "cleaner" way to do this, I have to pass all blocs everytime to every tab route, right? (or make it single bloc and pass this one instead).
This also was the reason why I wanted to somehow try that regex route, because there it would be passed cleanly only once.

**Reply by csells** [reply]
You bet. Check out [the navigator builder](https://pub.dev/packages/go_router#navigator-builder).

**Reply by leoshusar** [reply]
I was thinking of something I can use only for several routes (e.g. `/users`, `/roles` and not `/somethingelse`) but I think this would be enough.
Thank you for your help!

**Reply by DeD1rk** [reply]
You can of course move the providers into the TabScreen, but having them outside is basically the same if they are lazy.

**Reply by leoshusar** [reply]
Thanks, that's one way too. But I'd like to keep everything consistent and have all my providers in one place.

**Comment by csells** [comment]
The books example does just what you're asking for a believe

**Reply by leoshusar** [reply]
The books example uses regex for tab paths and all of those tabs have single `BookDetailsScreen`. So even when you're on `/books/popular` and open some book info, it goes to `/books/all/0`.
But I need:
- when I'm on `/roles` to open `RoleInfoScreen`
- when I'm on `/users`, I need to open `UserInfoScreen`.

I believe #133 would help with this. It would allow to write something like this:
```dart
final router = GoRouter(
  navigatorBuilder: (context, child, state) {
    if (RegExp('^/(books|roles)').hasMatch(state.subloc)) {
      return MultiProvider(
        providers: [
          Provider<UsersBloc>(...),
          Provider<RolesBloc>(...),
        ],
      child: child,
    }
  }
)
```

Another idea I got now is to have optional `routesBuilder` in `GoRoute` (allowed only `routes` or `routesBuilder`).
```dart
GoRoute(
  path: '/:main(users|roles)',
  pageBuilder: (c, s) => MaterialPage<void>(
    child: MultiProvider(
      providers: [
        Provider<UsersBloc>(...),
        Provider<RolesBloc>(...),
      ],
      child: TabsScreen(),
    ),
  ),
  routesBuilder: (state) {
    if (state.subloc.startsWith('/users'))
      GoRoute( 
        path: ':userId',
        pageBuilder: ... UserInfoScreen(),
      );
    if (state.subloc.startsWith('/roles'))
      GoRoute(
        path: ':roleId',
        pageBuilder: ... RoleInfoScreen(),
      );
  },
)
```
